src/controllers/EvidenceDownloadController.ts

```diff
@@ -0,0 +1,44 @@
+import { inject } from 'inversify';
+import { controller, httpGet, requestParam } from 'inversify-express-utils';
+import { Readable, Writable } from 'stream';
+import { BaseAsyncHttpController } from './BaseAsyncHttpController';
+
+import { FileTransferFeatureMiddleware } from 'app/middleware/FileTransferFeatureMiddleware';
+import { FileTransferService } from 'app/service/FileTransferService';
+import { DOWNLOAD_FILE_PAGE_URI } from 'app/utils/Paths';
+const template = 'download-file';
+
+@controller(DOWNLOAD_FILE_PAGE_URI, FileTransferFeatureMiddleware)
+export class EvidenceDownloadController extends BaseAsyncHttpController {
+
+    constructor(@inject(FileTransferService) private readonly fileTransferService: FileTransferService) {
+        super();
+    }
+
+    @httpGet('/prompt/:fileId')
+    public renderPrompt(@requestParam('fileId') fileId: string): void {
+        this.httpContext.response.render(template, { fileId });
+    }
+
+    @httpGet('/data/:fileId/download')
+    public async download(@requestParam('fileId') fileId: string): Promise<any> {
+
+        const res = this.httpContext.response;
+
+        const metadata = await this.fileTransferService.getFileMetadata(fileId);
+        res.setHeader('content-disposition', `attachment; filename=${metadata.name}`);
+
+        const readable = await this.fileTransferService.download(fileId);
+
+        return await this.pipeDataIntoStream(readable, res);
+
+    }
+
+    private async pipeDataIntoStream(readable: Readable, writable: Writable): Promise<void> {
```

Comment:
Does piping have to be promisified? Wouldn't it be enough to call `readable.pipe(writable)` directly in download method?

Reply:
![image](https://user-images.githubusercontent.com/54070318/78648482-eeda8980-78b3-11ea-8f38-ddbd13cbc124.png)

Short answer, yes.

Reply:
Alright, I can swallow that sour pill and we can improve it later ;)